swap/chain/txqueue.go

```diff
@@ -0,0 +1,639 @@
+}
+
+// stopWithError sends the error to the error channel
+// this is used to stop the queue from notification handlers
+func (txq *TxQueue) stopWithError(err error) {
+	select {
+	case txq.errorChan <- err:
+	default:
+		log.Error("failed to write error to txqueue error channel", "error", err)
+	}
+}
+
+// ScheduleRequest adds a new request to be processed
+// The request is assigned an id which is returned
+func (txq *TxQueue) ScheduleRequest(handlerID string, request TxRequest, extraData interface{}) (id uint64, err error) {
+	txq.lock.Lock()
+	defer txq.lock.Unlock()
+
+	// get the last id
+	err = txq.store.Get(txq.idKey(), &id)
+	if err != nil && err != state.ErrNotFound {
+		return 0, err
+	}
+	// ids start at 1
+	id++
+
+	// in a single batch this
+	// * stores the request data
+	// * stores the request extraData
+	// * adds it to the queue
+	batch := new(state.StoreBatch)
+	err = batch.Put(txq.idKey(), id)
+	if err != nil {
+		return 0, err
+	}
+
+	err = batch.Put(txq.extraDataKey(id), extraData)
+	if err != nil {
+		return 0, err
+	}
+
+	err = batch.Put(txq.requestKey(id), &txRequestData{
+		ID:        id,
+		Request:   request,
+		HandlerID: handlerID,
+		State:     TxRequestStateScheduled,
+	})
+	if err != nil {
+		return 0, err
+	}
+
+	_, triggerQueue, err := txq.requestQueue.Queue(batch, id)
+	if err != nil {
+		return 0, err
+	}
+
+	// persist to disk
+	err = txq.store.WriteBatch(batch)
+	if err != nil {
+		return 0, err
+	}
+
+	triggerQueue()
+	return id, nil
+}
+
+// GetExtraData load the serialized extra data for this request from disk and tries to decode it
+func (txq *TxQueue) GetExtraData(id uint64, request interface{}) error {
+	return txq.store.Get(txq.extraDataKey(id), &request)
+}
+
+// GetRequestState gets the state the request is currently in
+func (txq *TxQueue) GetRequestState(id uint64) (TxRequestState, error) {
+	var requestMetadata *txRequestData
+	err := txq.store.Get(txq.requestKey(id), &requestMetadata)
+	if err != nil {
+		return 0, err
+	}
+	return requestMetadata.State, nil
+}
+
+// Start starts processing transactions if it is not already doing so
+func (txq *TxQueue) Start() {
+	txq.lock.Lock()
+	defer txq.lock.Unlock()
+
+	if txq.started {
+		return
+	}
+
+	txq.started = true
+	txq.wg.Add(2)
+	go func() {
+		defer txq.wg.Done()
+		// run the actual loop
+		err := txq.loop()
+		if err != nil && !errors.Is(err, context.Canceled) {
+			log.Error("transaction queue terminated with an error", "queue", txq.prefix, "error", err)
+		}
+	}()
+
+	go func() {
+		defer txq.wg.Done()
+		// listen on the error channel and stop the queue on error
+		select {
+		case err := <-txq.errorChan:
+			log.Error("unrecoverable transaction queue error (transaction processing disabled)", "error", err)
+			txq.Stop()
+		case <-txq.ctx.Done():
+		}
+	}()
+}
+
+// Stop stops processing transactions if it is running
+// It will block until processing has terminated
+func (txq *TxQueue) Stop() {
```

Comment:
That's because they are passed in externally. Whichever component created those is response for closing them (in the case of production, this happens in `Swap`s `Close`).

Reply:
Got it!